collect_app/src/main/java/org/odk/collect/android/widgets/GeoPointWidget.java

```diff
@@ -17,153 +17,111 @@
+    public GeoPointWidget(Context context, QuestionDetails questionDetails, QuestionDef questionDef, WaitingForDataRegistry waitingForDataRegistry) {
         super(context, questionDetails);
-        determineMapProperties();
-    }
+        this.waitingForDataRegistry = waitingForDataRegistry;
+        accuracyThreshold = GeoWidgetUtils.getAccuracyThreshold(questionDef);
 
-    private void determineMapProperties() {
-        // Determine the accuracy threshold to use.
-        String acc = getFormEntryPrompt().getQuestion().getAdditionalAttribute(null, ACCURACY_THRESHOLD);
-        accuracyThreshold = acc != null && !acc.isEmpty() ? Double.parseDouble(acc) : DEFAULT_LOCATION_ACCURACY;
-
-        // Determine whether to use the map and whether the point should be draggable.
-        if (MapProvider.getConfigurator().isAvailable(getContext())) {
-            if (hasAppearance(getFormEntryPrompt(), PLACEMENT_MAP)) {
-                draggable = true;
-                useMap = true;
-            } else if (hasAppearance(getFormEntryPrompt(), MAPS)) {
-                draggable = false;
-                useMap = true;
-            }
+        stringAnswer = getFormEntryPrompt().getAnswerText();
+        boolean dataAvailable = false;
+        if (stringAnswer != null && !stringAnswer.isEmpty()) {
+            dataAvailable = true;
+            setBinaryData(stringAnswer);
         }
+        updateButtonLabelsAndVisibility(dataAvailable);
     }
 
-    public void updateButtonLabelsAndVisibility(boolean dataAvailable) {
-        if (useMap) {
-            if (readOnly) {
-                startGeoButton.setText(R.string.geopoint_view_read_only);
-            } else {
-                startGeoButton.setText(
-                    dataAvailable ? R.string.view_change_location : R.string.get_point);
-            }
+    @Override
+    protected View onCreateAnswerView(Context context, FormEntryPrompt prompt, int answerFontSize) {
+        binding = GeoWidgetAnswerBinding.inflate(((Activity) context).getLayoutInflater());
+        View answerView = binding.getRoot();
+
+        binding.geoAnswerText.setTextSize(TypedValue.COMPLEX_UNIT_DIP, answerFontSize);
+
+        readOnly = prompt.isReadOnly();
+        if (readOnly) {
+            binding.simpleButton.setVisibility(GONE);
         } else {
-            if (!readOnly) {
-                startGeoButton.setText(
-                    dataAvailable ? R.string.change_location : R.string.get_point);
-            }
+            binding.simpleButton.setText(getDefaultButtonLabel());
+            binding.simpleButton.setTextSize(TypedValue.COMPLEX_UNIT_DIP, answerFontSize);
+
+            binding.simpleButton.setOnClickListener(v -> {
+                Bundle bundle = GeoWidgetUtils.getGeoPointBundle(stringAnswer, accuracyThreshold, null, null);
+                GeoWidgetUtils.onButtonClick(context, prompt, getPermissionUtils(), null,
```

Comment:
For all the four GeoWidgets, I call static method `onButtonClick`, when `geoButton` is clicked. In the corresponding test, I was thinking just verifying the method call, and checking different cases later on in unit tests for `GeoWidgetUtils`.

Reply:
I think given you want to test the shared behaviour it's good to try and think hard about what the responsibility of `onButtonClick` really is here. For me what it does is ask for geo data for the widget. It feels to me like there could be an interface called something like `GeoDataRequester` or `GeoDataFetcher` with methods like `requestPoint`, `requestShape` etc:

```
binding.simpleButton.setOnClickListener(v -> {
    geoDataRequester.requestPoint(promp);
});
```

 In each of the different geo widgets you can have a test that the `requestPoint` or whatever is called and then those methods can be tested for some implementation of the interface (`ActivityGeoDataRequester`).

Just to detail my questions here the though process I'm going through is:

1. I have shared code used in multiple places does it make sense to just test the behaviour in each place?
2. If not, what is the shared code responsible for? Is there some kind of interface/object/entity that could be responsible for that?